File: migalha/MultiLabel.java
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

// combines multiple Labels in order to analyze many vulnerabilities at once, keeping their flows separate and independent
public class MultiLabel {
    private final Map<Pattern, Label> labels;

    public MultiLabel() {
        this.labels = new HashMap<>();
    }

    public MultiLabel(Set<Pattern> patterns) {
        this.labels = new HashMap<>();
        for (Pattern p : patterns) {
            this.labels.put(p, new Label());
        }
    }

    public MultiLabel(MultiLabel other) {
        this.labels = new HashMap<>();
        for (Map.Entry<Pattern, Label> entry : other.labels.entrySet()) {
            this.labels.put(entry.getKey(), new Label(entry.getValue())); 
        }
    }

    public Map<Pattern, Label> getLabels() {
        return Collections.unmodifiableMap(labels);
    }

    public Label getLabel(Pattern p) {
        return labels.get(p);
    }

    public void addSource(String sourceName) {
        for (Pattern p : labels.keySet()) {
            if (p.isSource(sourceName)) {
                labels.get(p).addSource(sourceName);
            }
        }
    }

    public void addSanitizer(String sourceName, String sanitizerName) {
        for (Pattern p : labels.keySet()) {
            if (p.isSanitizer(sanitizerName)) {
                if (p.isSource(sourceName)) {
                    labels.get(p).addSanitizer(sourceName, sanitizerName);
                }
            }
        }
    }

    public MultiLabel combine(MultiLabel other) {
        MultiLabel combined = new MultiLabel(this);

        for (Pattern p : other.labels.keySet()) {
            Label otherLabel = other.labels.get(p);

            combined.labels.putIfAbsent(p, new Label());

            Label merged = combined.labels.get(p).combine(otherLabel);
            combined.labels.put(p, merged);
        }

        return combined;
    }
}

File: migalha/MultiLabel.py
from typing import Dict, Set, Mapping
from copy import deepcopy
from Label import Label

class MultiLabel:
    """
    Combines multiple Labels in order to analyze many vulnerabilities at once,
    keeping their flows separate and independent.
    """

    def __init__(self, patterns: Set = None, other: "MultiLabel" = None):
        if other is not None:
            # Copy constructor
            self._labels: Dict = {
                p: Label(label) for p, label in other._labels.items()
            }
        elif patterns is not None:
            # Init from a set of patterns
            self._labels = {p: Label() for p in patterns}
        else:
            # Empty constructor
            self._labels = {}

    def get_labels(self) -> Mapping:
        """
        Returns an immutable view similar to Java's Collections.unmodifiableMap.
        """
        # Shallow immutability: keys → patterns (unchanged), values → Label instances (mutable).
        # This matches Java semantics.
        return dict(self._labels)

    def get_label(self, pattern):
        return self._labels.get(pattern)

    def add_source(self, source_name: str) -> None:
        for p in self._labels:
            if p.isSource(source_name):
                self._labels[p].add_source(source_name)

    def add_sanitizer(self, source_name: str, sanitizer_name: str) -> None:
        for p in self._labels:
            if p.isSanitizer(sanitizer_name) and p.isSource(source_name):
                self._labels[p].add_sanitizer(source_name, sanitizer_name)

    def combine(self, other: "MultiLabel") -> "MultiLabel":
        combined = MultiLabel(other=self)  # copy of self

        for p, other_label in other._labels.items():
            if p not in combined._labels:
                combined._labels[p] = Label()

            combined._labels[p] = combined._labels[p].combine(other_label)

        return combined

    def __repr__(self):
        return f"MultiLabel({self._labels})"
